### PR TITLE
Toggl Track - Added Features

### PR DESCRIPTION
**Pull request by Gomeyy**
## Description

- Action to create a complete time entry with a start and stop time.
- Action to delete/edit time entries (included a preference to turn on deletes to stop accidental deletes).
- Ability to set a default project in preferences (optional)

## Type of change

- Bug fix / improvement for somebody else's extension. Extension author: @franzwilhelm #734

## Screencast

image
image
image
image
image
image

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
	Was unable to resolve two errors. Both are because I’m setting date pickers using a string instead of date which was the work around I found to be able to set a default date and time.

### CODE REVIEW

**Comment by pernielsentikaer** [comment]
Hi @Gomeyy 👋 

Could you add a label attribute to your preferences to please the linter 💪 😄 

@franzwilhelm can you maybe test it out and give some feedback? 🔥

File: extensions/toggl-track/src/components/EditTimeEntryForm.tsx [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+      setSelectedProject(project);
+    }
+  };
+
+  const onTagsChange = (tags: string[]) => {
+    setSelectedTags(tags);
+  };
+
+  return(
+    <Form
```
Comment (franzwilhelm) [inline extensions/toggl-track/src/components/EditTimeEntryForm.tsx:59]: This form should be extracted to its own component, so the interaction is the same for creating time entries and editing time entries

File: extensions/toggl-track/src/components/EditTimeEntryForm.tsx [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                key={project.id}
+                value={project.id.toString()}
+                title={project.name}
+                icon={{ source: Icon.Circle, tintColor: project.hex_color }}
+              />
+            ))}
+          </Form.Dropdown.Section>
+        ))}
+      </Form.Dropdown>
+      <Form.TagPicker id="tags" title="Tags" onChange={onTagsChange} defaultValue={entry.tags}>
```
Comment (franzwilhelm) [inline extensions/toggl-track/src/components/EditTimeEntryForm.tsx:94]:
```suggestion
      <Form.TagPicker id="tags" title="Tags" onChange={setSelectedTags} defaultValue={entry.tags}>
```

File: extensions/toggl-track/src/components/EditTimeEntryForm.tsx [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  const onProjectChange = (projectId: string) => {
+    const project = projects.find((project) => project.id === parseInt(projectId));
+    if (project) {
+      setSelectedProject(project);
+    }
+  };
+
+  const onTagsChange = (tags: string[]) => {
+    setSelectedTags(tags);
+  };
```
Comment (franzwilhelm) [inline extensions/toggl-track/src/components/EditTimeEntryForm.tsx:56]:
```suggestion
```

File: extensions/toggl-track/src/components/EditTimeEntryForm.tsx [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import { useAppContext } from "../context";
+import { useMemo, useState } from "react";
+
+function EditTimeEntryForm({ project, entry }: { project?: Project; entry: TimeEntry }) {
+  const navigation = useNavigation();
+  const { projects, tags, isLoading, projectGroups } = useAppContext();
+  const [selectedProject, setSelectedProject] = useState<Project | undefined>(project);
+  const [selectedTags, setSelectedTags] = useState<string[]>([]);
+
+  async function handleSubmit(values: { description: string; tags: string[]; start: string; end?: string}) {
```
Comment (franzwilhelm) [inline extensions/toggl-track/src/components/EditTimeEntryForm.tsx:23]: No need to keep a selectedProject state, as we can get the project id from values

Comment (franzwilhelm) [reply extensions/toggl-track/src/components/EditTimeEntryForm.tsx:23]: Same goes for tags

File: extensions/toggl-track/src/components/EditTimeEntryForm.tsx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+  ToastStyle,
+  clearSearchBar,
+} from "@raycast/api";
+import toggl from "../toggl";
+import { storage } from "../storage";
+import { Project, TimeEntry } from "../toggl/types";
+import { useAppContext } from "../context";
+import { useMemo, useState } from "react";
+
+function EditTimeEntryForm({ project, entry }: { project?: Project; entry: TimeEntry }) {
```
Comment (franzwilhelm) [inline extensions/toggl-track/src/components/EditTimeEntryForm.tsx:17]: No need to pass in project. Can be inferred from the time entry's pid

File: extensions/toggl-track/src/components/CreateTimeEntryForm.tsx [hunk 6]
```diff
@@ -7,33 +7,42 @@ import {
...
 
-  async function handleSubmit(values: { description: string }) {
+  async function handleSubmit(values: { description: string; start: string; end: string; }) {
     try {
       await toggl.createTimeEntry({
         projectId: selectedProject?.id,
         description: values.description,
         tags: selectedTags,
+        start: values.start,
+        duration: Math.trunc((new Date(values.end).getTime() - new Date(values.start).getTime()) / 1000)
```
Comment (franzwilhelm) [inline extensions/toggl-track/src/components/CreateTimeEntryForm.tsx:37]: Doesn't the Toggl Track API passing the end time directly as `end`? This is safer than calculating the duration. Pass in `end: values.end` instead?

File: extensions/toggl-track/src/index.tsx [hunk 7]
```diff
@@ -11,31 +11,40 @@ import {
...
 
 dayjs.extend(duration);
 
 function ListView() {
   const { isLoading, isValidToken, projectGroups, runningTimeEntry, timeEntries, projects } = useAppContext();
   const getProjectById = (id: number) => projects.find((p) => p.id === id);
-
   const timeEntriesWithUniqueProjectAndDescription = timeEntries.reduce((acc, timeEntry) => {
-    const existing = acc.find((t) => t.description === timeEntry.description && t.pid === timeEntry.pid);
+    const existing = acc.find((t) => (t.description === timeEntry.description && t.pid === timeEntry.pid) || !("stop" in timeEntry));
```
Comment (franzwilhelm) [inline extensions/toggl-track/src/index.tsx:36]: Not sure what `"stop" in timeEntry` solves here?

File: extensions/toggl-track/src/index.tsx [hunk 8]
```diff
@@ -11,31 +11,40 @@ import {
...
     if (!existing) {
       acc.push(timeEntry);
     }
     return acc;
   }, [] as TimeEntry[]);
+  timeEntriesWithUniqueProjectAndDescription.reverse()
```
Comment (franzwilhelm) [inline extensions/toggl-track/src/index.tsx:42]:
```suggestion
```

Use `reduceRight` above instead

File: extensions/toggl-track/src/index.tsx [hunk 9]
```diff
@@ -90,6 +122,22 @@ function ListView() {
                           onSubmit={() => resumeTimeEntry(timeEntry)}
                           icon={{ source: Icon.Clock }}
                         />
+                        <PushAction
```
Comment (franzwilhelm) [inline extensions/toggl-track/src/index.tsx:125]: I really love that editing/deleting the current running time entry has been added. But I don't think it makes sense to be able to delete/edit previous entries as the extension is set up currently. The entries shown in the list are not representative of _all_ entries, but only unique groups of name+pid. In my workflow I often create multiple time entries with the same name and project id. So in my case, which time entry exactly is deleted? I suggest to remove the ability to delete and edit previous time entries, but keep the possibility to edit and delete the running time entry.

We could try to make the list of time entries more like on the Toggl Track web edition. But as of now this extension is not ready for edit/delete of time entry history

File: extensions/toggl-track/package.json [hunk 10]
```diff
@@ -23,6 +23,21 @@
...
+    {
+      "name": "defaultProject",
+      "type": "textfield",
+      "required": false,
+      "title": "Default Project ID",
+      "placeholder": "Can be found in the URL when viewing a project.",
+      "description": "The default project when starting/creating a time entry."
+    },
+    {
+      "name": "allowDelete",
```
Comment (franzwilhelm) [inline extensions/toggl-track/package.json:36]: Ref comments below. This config variable can be removed

File: extensions/toggl-track/src/components/RunningTimeEntry.tsx [hunk 11]
```diff
@@ -33,6 +35,15 @@ function RunningTimeEntry({ runningTimeEntry }: { runningTimeEntry: TimeEntry })
         actions={
           <ActionPanel>
             <SubmitFormAction icon={{ source: Icon.Clock }} onSubmit={stopTimeEntry} title="Stop Time Entry" />
+            <PushAction
+              title="Edit Time Entry"
+              icon={{ source: Icon.Clock }}
+              target={
```
Comment (franzwilhelm) [inline extensions/toggl-track/src/components/RunningTimeEntry.tsx:41]: Add delete time entry here? Could be cool to "cancel" the running time entry :D

**Review by franzwilhelm — CHANGES_REQUESTED** [review]
Great additions @Gomeyy ! Thanks for improving the extension. I do have some questions and comments that should be resolved before considering this addition ready. 

Looking forward to your response :D If you need help with any of the suggestions, just tell me and I'll see what I can do

**Comment by Gomeyy** [comment]
@franzwilhelm Thanks for the input, I will work on this when I get some free time!

File: extensions/toggl-track/src/index.tsx (hunk 7 above)
Comment (Gomeyy) [reply extensions/toggl-track/src/index.tsx:36]: It stops the current running time entry from showing up under recent time entries, so it doesn't appear twice.

File: extensions/toggl-track/src/index.tsx (hunk 9 above)
Comment (Gomeyy) [reply extensions/toggl-track/src/index.tsx:125]: I agree, It currently deletes the most recent time entry of the group, which has been useful for me for some quick deletes, but I've also had to go onto the site to delete things at times.

**Comment by pernielsentikaer** [comment]
Wow, thanks a bunch for the review @franzwilhelm 🔥 Much appreciated 💪

**Comment by fe9lix** [comment]
Hey @Gomeyy, could you give an ETA when you're able to continue working on this? Btw, the CI seems to be failing currently due to some preference validation errors.

**Comment by franzwilhelm** [comment]
@fe9lix I should be able to fix the issues addressed in my review this weekend. :)

**Comment by sxn** [comment]
hi @franzwilhelm @Gomeyy - any updates on this? There's been no activity for almost a month. Should we close the PR or set it to draft mode until you manage to implement the changes? 😸

**Comment by franzwilhelm** [comment]
@Gomeyy great to see you working on this again 🤝 Could you resolve my review comments when fixed? Also don't hesitate to ask for help / feedback. I won't have time to work actively on this, but I can always assist a bit.

@sxn if @Gomeyy has the time to fix the issues we can keep the PR open. If not, we can draft this in a week.

**Comment by pernielsentikaer** [comment]
Drafting this extension, feel free to open it again when the time is there 😄

**Comment by pernielsentikaer** [comment]
@franzwilhelm, what do you think it takes to prepare this for merge? I would be sad to close the PR now that it feels like it's almost prepared 😄

**Comment by pernielsentikaer** [comment]
Closing this due to inactivity, feel free to open it again when you're ready to continue working on it 😄

**Pull request closed without merging**